modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentFailureTest.java

```diff
@@ -0,0 +1,231 @@
+
+        IgniteEx srv = startGrid(getConfiguration("server"));
+        IgniteEx cli = startClientGrid(1);
+
+        ServiceConfiguration svcCfg = new ServiceConfiguration()
+                .setName(srvcName)
+                .setService(((Class<Service>)extClsLdr.loadClass(NOOP_SERVICE_CLS_NAME)).getDeclaredConstructor().newInstance())
+                .setNodeFilter(((Class<IgnitePredicate<ClusterNode>>)extClsLdr.loadClass(NODE_FILTER_CLS_NAME))
+                        .getConstructor(UUID.class)
+                        .newInstance(cli.configuration().getNodeId()))
+                .setTotalCount(1);
+
+        assertThrowsWithCause(() -> cli.services().deploy(svcCfg), ServiceDeploymentException.class);
+
+        assertTrue(waitForCondition(() -> noDescriptorInClusterForService(srvcName), TIMEOUT));
+    }
+
+    /**
+     * Tests that service descriptors are clear after an attempt of deploying
+     * a service which throws an exception during initialization
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testServerDescriptorsOfFailedServices() throws Exception {
+        final int noopSrvcTotalCnt = 20;
+        final int noopSrvcMaxPerNodeCnt0 = 2;
+        final int noopSrvcMaxPerNodeCnt1 = 4;
+
+        final int throwSrvcTotalCnt = 10;
+        final int throwSrvcMaxPerNodeCnt = 2;
+
+        startGrids(SERVER_NODES_CNT);
+
+        IgniteEx client = startClientGrid(SERVER_NODES_CNT);
+
+        for (int i = 1; i < CLIENT_NODES_CNT; i++)
+            startClientGrid(SERVER_NODES_CNT + i);
+
+        ServiceConfiguration noopSrvcCfg = new ServiceConfiguration()
+                .setName(NoopService.class.getSimpleName())
+                .setService(new NoopService())
+                .setTotalCount(noopSrvcTotalCnt)
+                .setMaxPerNodeCount(noopSrvcMaxPerNodeCnt0);
+
+        // Deploying NoopService - should be completed successfully.
+        client.services().deploy(noopSrvcCfg);
+
+        // Check that the expected number of instances has been deployed
+        assertEquals(noopSrvcMaxPerNodeCnt0 * SERVER_NODES_CNT,
+                totalInstancesCount(client, NoopService.class.getSimpleName()));
+
+        // Deploy InitThrowingService that throws an exception in init - should not be deployed.
+        assertThrowsWithCause(() -> client.services().deploy(new ServiceConfiguration()
+                        .setName(InitThrowingService.class.getSimpleName())
+                        .setService(new InitThrowingService())
+                        .setTotalCount(throwSrvcTotalCnt)
+                        .setMaxPerNodeCount(throwSrvcMaxPerNodeCnt)),
+                ServiceDeploymentException.class);
+
+        // Wait until the descriptors are updated on all nodes and check that there are no descriptors of an undeployed service.
+        assertTrue(waitForCondition(() -> noDescriptorInClusterForService(InitThrowingService.class.getSimpleName()), TIMEOUT));
```

Comment:
Let's add check that NoopService that was deployed successfully present in descriptors on all nodes.